text/0000-structural-records.md

```diff
@@ -0,0 +1,1298 @@
+```
+
+we cannot know whether it will become a block with a type ascription,
+i.e. `{ $ident: $type }`, or if it will become a structural record
+`{ $ident: $expr, }` (the comma makes this unambiguous).
+
+[RFC 2544]: https://github.com/rust-lang/rfcs/pull/2544
+
+This changes the formal parsing complexity of Rust such that it is no
+longer LL(k). While this is a nice property to retain, this is quite a
+localized instance of backtracking that a parser combinator library
+should be able to deal with easily. See [RFC 2544] for a discussion.
+
+There are also other places where allowing backtracking could be
+useful such as with [RFC 2544]. Furthermore, since `{ $ident: $type }`
+is not likely to occur frequently in the wild even if type ascription
+is stabilized, performance is unlikely to suffer notably.
+
+Another instance where backtracking / lookahead may be required is when parsing:
+
+```rust
+fn f() where $constraint_1, { $ident:
+```
+
+In this case, there is no ambiguity, but `{ $ident: ...` could either be the
+start of a structural record type in the second constraint or belonging to the
+function body (type ascription on an identifier). The second case is likely
+pathological, and so it is unlikely to occur much in practice.
+
+## Hard to implement crate-external traits
+
+As with positional tuples, because a structural record is never crate local,
+this presents users with a problem when they need to implement a trait they
+don't own for a structural record comprising of crate-local types.
+
+For example, say that you have the crate-local types `Foo` and `Bar`.
+Both of these types implement `serde::Serialize`.
+Now you'd like to serialize `type T = { foo: Foo, bar: Bar };`.
+However, because neither `serde::Serialize` nor `T` is crate-local,
+you cannot `impl serde::Serialize for T { ... }`.
+
+This inability is both a good and a bad thing. The good part of it is that it
+might prevent overuse of structural records and provide some pressure towards
+nominal typing that might be good for robustness. The bad part is that these
+sort of one-off structures are a good reason to have structural records in the
+first place. With some combined quantification of field labels (possibly via
+const generics), and with tuple-variadic generics, it should be possible (for
+`serde`, if there is a will, to offer implementations of `Serialize` for all
+structural records.
+
+Note that while `impl serde::Serialize for T { ... }` may not be possible
+without extensions, the following would be:
+
+```rust
+#[derive(Serialize)]
+struct RectangleTidy {
+    dimensions: {
+        width: u64,
+        height: u64,
+    },
+    color: {
+        red: u8,
+        green: u8,
+        blue: u8,
+    },
+}
+```
+
+## "Auto-implementing traits is a magical hack"
+
+Indeed, we would much prefer to use a less magical approach,
+but providing these traits without compiler magic would require
+significantly more complexity such as polymorphism over field names
```

Comment:
I think you underestimate the compiler complexity needed to support this as written in the RFC. Additionally this will cause not insignificant code bloat for all the impls required.

I think the most reasonable way to support this RFC, plus loads of default trait impls, plus custom trait impls for users is to desugar these types (and their field accesses) to tuples (and tuple field accesses).

From the user perspective, this means that if you implement a trait for `(T, U)`, then it is also available for `{bar: T, foo: U}` assuming that `foo` is sorted after `bar`. It is left to the one who implements a trait for tuples to ensure that the impls are order independent.

From the compiler perspective, we'd create a new `TyKind`:

```rust
    /// The second field is equivalent to the field for the `Tuple` variant
    /// The first field is a list of names for the tuple's fields
    StructuralRecord(&[Name], &[Ty<'tcx>]),
```

(or we just add this first field to `Tuple` in general (filling it in with `["0", "1", ...]` for normal tuples and do everything below for the `Tuple` variant)

At this point we guarantee that the memory layout of structural records is equivalent to that of their corresponding tuple. This means the compiler may transmute freely between the two. So when we're calling `<{bar: u32, foo: String} as PartialEq>::eq` will actually invoke `<(u32, String) as PartialEq>::eq` by effectively transmuting each argument (`&self` and `other: &{bar: u32, foo: String}`) to `&(u32, String)`

Field accesses via `.foo` or `.bar` can be lowered to `.1` and `.0` respectively after typechecking.

The only downside I see with this is that `<{bar: u32, foo: String} as Debug>::fmt` will print a tuple and not the fields and that `<{bar: u32, foo: String} as PartialEq<(u32, String)>` actually typechecks.

Also for `serde`, all structural records would be serialized as tuples, losing the advantage of named fields in human readable serialization formats.

Reply:
I'd agree the code works exactly like tuples, and code deduplicating rocks, but we need structural records to be different types from tuples if we want this to serve much purpose.

Reply:
With this scheme they'd be different types up to the point of trait resolution. So you can't pass a `(u32, String)` where a `{bar: u32, foo: String}` is expected and vice versa. But when you use trait methods, then they are the same type.

Reply:
I'm not sure this is a good idea. Things like comparisons have lexicographic ordering, which would observe an order we don't want to exist.

Also, implementing something like field-name/row polymorphism isn't as hard as coming up with a syntax for it.

Reply:
Note that `Copy` & `Clone` are already automatically generated by `rustc` for tuples and closures *bypassing `derive`*, so I don't think extending anything that works for closures to records is fundamentally problematic in any way.

@Centril basically, feel free to rely on #2132 and #2133 as precedent.
*However*, note that `Default` is not included, so you'd have to probably make an RFC to add compiler-provided `Default` to tuples, if not closures.

Reply:
I agree about `serde`. I don't think there's a clean solution without <https://github.com/rust-lang/rfcs/pull/2584#issuecomment-436279761>.

At this point, const generics and/or VG are required for experimenting with any kind of anonymous record scheme, that can actually handle `serde`, if we don't want to outright add subtyping(-like) constructs to the language.

Reply:
@oli-obk 

> I think you underestimate the compiler complexity needed to support this as written in the RFC. Additionally this will cause not insignificant code bloat for all the impls required.

I don't think I've estimated the compiler complexity at all; I've provided a specification of the behavior, not exactly how it would be implemented. I fully expect some complexity to come from this. There's an infinite number of implementations possible; so they will need to be lazily provided.
If de-duplication is the concern then some sort of de-duplication of impls could ostensibly be provided the fields are of no concern.

> From the user perspective, this means that if you implement a trait for `(T, U)`, then it is also available for `{bar: T, foo: U}` assuming that `foo` is sorted after `bar`. It is left to the one who implements a trait for tuples to ensure that the impls are order independent.

This would make that behavior observable in the type system such that `(T, U)` and `{bar: T, foo: U}` are effectively the same type.

> At this point we guarantee that the memory layout of structural records is equivalent to that of their corresponding tuple.

The RFC already states that this is the case (in the section on dynamic semantics) but equivalent memory layout is not a sufficient condition to permit free transmutation. I think this would need to be stated separately that the compiler may freely transmute them.

> The only downside I see with this is that `<{bar: u32, foo: String} as Debug>::fmt` will print a tuple and not the fields and that `<{bar: u32, foo: String} as PartialEq<(u32, String)>` actually typechecks.

This is quite the drawback and quite surprising... It would also mean that implementations are only provided for records of size up to 12. (pending VG). I think the behavior wrt. `PartialEq` this is fine as an implementation strategy as long as no behavior is observable 

> 
> Also for `serde`, all structural records would be serialized as tuples, losing the advantage of named fields in human readable serialization formats.

Also not a great drawback.

> At this point, const generics and/or VG are required for experimenting with any kind of anonymous record scheme, that can actually handle `serde`, if we don't want to outright add subtyping(-like) constructs to the language.

I don't think we want to add subtyping-like constructs (row polymorphism is not sub-typing so it doesn't count).

With respect to const generics and VG, I already noted this in https://github.com/rust-lang/rfcs/pull/2584#issuecomment-435556892. This is not unheard of... GHC allows you to sort of do this so that structural records can be provided well as a library solution.